### PR TITLE
[MM-40962-2] - Fix unnecessary logging in license expiration check

### PR DESCRIPTION
**Pull request by AGMETEOR**
#### Summary
This PR is a redo of https://github.com/mattermost/mattermost-server/pull/19653 and does the following
- Puts back the log that was removed after [this suggestion](https://github.com/mattermost/mattermost-server/pull/19653#issuecomment-1073680601)
- Skips check for TE
- Does a refactor to generate one renewal link for all admins when sending a "SendRemoveExpiredLicenseEmail". This means that in the case of an error in the renewal JWT token stored, we don't log this error for all admins present on the system

#### Ticket Link
[MM-40962](https://mattermost.atlassian.net/browse/MM-40962)

#### Release Note

```release-note
NONE
```

### CODE REVIEW

File: app/server.go
Comment (AGMETEOR) [inline app/server.go:1724]: Hey @agnivade do you have any idea why this was added in the first place?

File: app/server.go
Comment (AGMETEOR) [inline app/server.go:1772]: Hey @agnivade , this too

File: app/server.go
Comment (AGMETEOR) [reply app/server.go:1724]: This produces a weird effect due to a listener that was added, the listener then sends a websocket to the client with a sanitized license, what happens then is that, on the license page in the console, we have some details missing

File: app/server.go
Comment (marianunez) [inline app/server.go:1781]: What do we gain extracting this?

File: app/license.go
Comment (marianunez) [inline app/license.go:385]: I guess the question here comes down to:
1. Why do we store the token in the System Store? It seems that we want to generate a new token for a link on demand and here if there is an error validating - we ignore and generate a new one anyways.

File: app/server.go
Comment (agnivade) [reply app/server.go:1724]: I don't think I am aware of the context behind this. I think this part is owned by your team so I'll leave it to your best judgement.

File: app/license.go
Comment (agnivade) [reply app/license.go:385]: I don't mind either ways. Technically, it seems fine to return an existing token if it hasn't expired. I guess the main question we had was "Why are we getting a spam of these token renewal requests during load-tests"? Do we know why is that?

Comment (marianunez) [reply app/license.go:385]: +1, this should only be called if we are sending the license expired email or in the front end calling the endpoint to get renewal link. Are any of these actions happening in load testing?

Comment (AGMETEOR) [reply app/license.go:385]: @agnivade the spam of tokens was due an attempt to generate a token for each admin when sending the email. So if multiple admins exist then we get several of those logs.

@marianunez my thought also was we don’t need to store the token in the db. We can always have one generated on the fly

File: app/server.go
Comment (AGMETEOR) [reply app/server.go:1781]: In case of logs having in that function call, it doesn’t log for each admin

File: app/license.go
Comment (marianunez) [reply app/license.go:385]: >  the spam of tokens was due an attempt to generate a token for each admin when sending the email. So if multiple admins exist then we get several of those logs.

Why is the load test triggering these emails?

Comment (isacikgoz) [reply app/license.go:385]: I think it was getting triggered by `ExpiryNotify` job.

Comment (AGMETEOR) [reply app/license.go:385]: @isacikgoz not sure I understand how it does. We trigger this job as a recurring task

Comment (agnivade) [reply app/license.go:385]: We need to get to the bottom of this as this is blocking the release. @isacikgoz - The question I had is do we see the logs being repeated or is it a one time spam of logs and it stops? I am wondering if we can create a separate Terraform instance and hand it off to @AGMETEOR to debug it further?

I guess you could not reproduce this locally right?

Comment (AGMETEOR) [reply app/license.go:385]: Yes, I couldn't reproduce it locally.

Comment (isacikgoz) [reply app/license.go:385]: @AGMETEOR actually you can first try the load test agent on your local instance, maybe we wouldn't need to go over the aws path? WDYT?

Comment (agnivade) [reply app/license.go:385]: Oh, can we reproduce this with a local load test? Then it should be fairly easy to debug.

File: app/server.go
Comment (marianunez) [reply app/server.go:1772]: Same here https://github.com/mattermost/mattermost-server/pull/19826#discussion_r836438453

File: app/server.go
Comment (agnivade) [reply app/server.go:1724]: @AGMETEOR - do you want to bring this back or are you confident this is good to be removed?

File: app/license.go
Comment (agnivade) [reply app/license.go:385]: @AGMETEOR / @isacikgoz -  Do we want to just move on with the review of the PR given that we weren't able to reproduce this a second time?

Comment (AGMETEOR) [reply app/license.go:385]: Yes.

File: app/server.go
Comment (agnivade) [inline app/server.go:1729]: expiration

File: app/server.go
Comment (agnivade) [inline app/server.go:1760]: Hmm .. any reason to invert the check? This creates more indented code. In Go, we prefer early returns.

File: app/server.go
Comment (AGMETEOR) [reply app/server.go:1724]: I am 1/5 removing it

**Review by agnivade — CHANGES_REQUESTED** [review]

File: app/server.go
Comment (AGMETEOR) [reply app/server.go:1760]: Screenshot 2022-03-30 at 18 43 24
I wonder why github is not reflecting that commit

Comment (AGMETEOR) [reply app/server.go]: my bad I had not pushed

File: app/license.go
Comment (marianunez) [inline app/license.go:385]: Looking more into the errors reported in the load testing `token is expired by`, this is actually returned from the [JWT library here](https://github.com/mattermost/mattermost-server/blob/baf9f69c273c67d8c1569c6bab9376d21971d8cf/vendor/github.com/dgrijalva/jwt-go/claims.go#L40). Which means that we are reporting an error for the token being expired which we know it will eventually happen given that it has a time period that it is valid and we are saving it for some reason. 

Again, I don't think there is any value on logging a warning for something that is expected and we should just be generating a new token each time - like I mentioned above. 

I'm 4/5 that we should **not** keep this error as is. We should either:

1. Stop storing the token in the System Store and just generate the token on demand. This would avoid storing something that will expire and produce these unnecessary errors/checks. 
2. Surface the error type [ValidationErrorExpired](https://github.com/mattermost/mattermost-server/blob/baf9f69c273c67d8c1569c6bab9376d21971d8cf/vendor/github.com/dgrijalva/jwt-go/claims.go#L41) from our `renewalTokenValid` and check here so we can ignore

**Review by marianunez — CHANGES_REQUESTED** [review]

File: app/license.go
Comment (agnivade) [reply app/license.go:385]: Fantastic analysis Maria! Yeah, fully agree we can ignore the expired errors.

**Review by agnivade — APPROVED** [review]
Good to go after Maria's comments are addressed

File: app/license.go
Comment (AGMETEOR) [reply app/license.go]: @marianunez went with option 1

**Review by marianunez — APPROVED** [review]
Thanks @AGMETEOR!

**Comment by AGMETEOR** [comment]
Got 2 dev reviews, so I am going to go ahead and merge

**Comment by marianunez** [comment]
> Got 2 dev reviews, so I am going to go ahead and merge

We should also do a quick smoke test on the renew now link and email to make sure there is no regression from these. O/5 if we want to do this post merge cc/@steve.mudie

**Comment by AGMETEOR** [comment]
Hey @stevemudie . Post merge?

**Comment by stevemudie** [comment]
> Hey @stevemudie . Post merge?

Yup, sounds good.

**Pull request merged**
